### PR TITLE
Open election support for backend 1

### PR DESCRIPTION
**Pull request by Robingoumaz**
Added a election#open message according to PR #657 and #850. 

Added a new callback method to handle the message in channel/election/mod.go
Added a verification method to check election#open messages in channel/election/verification.go
Changed the verification of other messages to accommodate the started and terminated state in channel/election/verification.go

### CODE REVIEW

**Comment by ahominal** [comment]
Right now, while processing the OpenElection message, it checks when the election is scheduled to start (even in the future). Wouldn't it be more logical to check current timestamp (election is opened now)?

File: be1-go/channel/election/mod_test.go [hunk 1]
```diff
@@ -340,6 +341,93 @@ func Test_Cast_Vote_And_Gather_Result(t *testing.T) {
...
+		WitnessSignatures: []message.WitnessSignature{},
+	}
+
+	// Fail to process non election open
+	require.Error(t, electChannel.processElectionOpen(m, messagedata.ElectionEnd.NewEmpty, socket.OrganizerSocket{}))
+
+	// Fail for non base64 sender key
+	require.Error(t, electChannel.processElectionOpen(m, electionOpen, socket.OrganizerSocket{}))
+
+	// Fail for not organizer public key
```
Comment (Warhedgehog) [inline be1-go/channel/election/mod_test.go:426]: Small typo here I believe

**Review by Warhedgehog — COMMENTED** [review]
Very good work 👍 A couple minor details where I'm unsure.

**Review by Warhedgehog — APPROVED** [review]
LGTM

File: be1-go/channel/election/mod.go [hunk 2]
```diff
@@ -25,6 +25,7 @@ import (
 
 const msgID = "msg id"
 const failedToDecodeData = "failed to decode message data: %v"
+const failedToBroadcast = "failed to broadcast message: %v"
```
Comment (MaximeZmt) [inline be1-go/channel/election/mod.go:28]:
```suggestion
const (
msgID = "msg id"
failedToDecodeData = "failed to decode message data: %v"
failedToBroadcast = "failed to broadcast message: %v"
)
```
What do you think of putting const like that ?

File: be1-go/channel/election/verification.go [hunk 3]
```diff
@@ -10,39 +10,102 @@ import (
 	"golang.org/x/xerrors"
 )
 
+const laoIDBase64 = "lao id is %s, should be base64URL encoded"
+const elecIDBase64 = "election id is %s, should be base64URL encoded"
+const elecIDFormat = "election channel id is %s, should be formatted as /root/laoID/electionID"
+const laoIDCompare = "lao id is %s, should be %s"
+const elecIDCompare = "election id is %s, should be %s"
```
Comment (MaximeZmt) [inline be1-go/channel/election/verification.go:17]:
```suggestion
const (
laoIDBase64 = "lao id is %s, should be base64URL encoded"
elecIDBase64 = "election id is %s, should be base64URL encoded"
elecIDFormat = "election channel id is %s, should be formatted as /root/laoID/electionID"
laoIDCompare = "lao id is %s, should be %s"
elecIDCompare = "election id is %s, should be %s"
)
```
same here ?

**Comment by MaximeZmt** [comment]
👍 LGTM (just two suggestions above, then it's a matter of personal opinion at that point 😂 )

File: be1-go/channel/election/mod.go [hunk 4]
```diff
@@ -297,6 +306,53 @@ func (c *Channel) verifyMessage(msg message.Message) error {
...
 }
 
+func (c *Channel) processElectionOpen(msg message.Message, msgData interface{}, _ socket.Socket) error {
+
+	_, ok := msgData.(*messagedata.ElectionOpen)
+	if !ok {
+		return xerrors.Errorf("message '%T' isn't a election#open message", msgData)
+	}
+
+	c.log.Info().Msg("received a election#open message")
```
Comment (nkcr) [inline be1-go/channel/election/mod.go:316]: Would be a good idea to add some context, like:

```go
c.log.Info().Str("Election", msgData.Election).Msg("received a election#open message")
```

File: be1-go/channel/election/mod.go [hunk 5]
```diff
@@ -297,6 +306,53 @@ func (c *Channel) verifyMessage(msg message.Message) error {
...
+
+	senderBuf, err := base64.URLEncoding.DecodeString(msg.Sender)
+	if err != nil {
+		return xerrors.Errorf("failed to decode sender key: %v", err)
+	}
+
+	senderPoint := crypto.Suite.Point()
```
Comment (nkcr) [inline be1-go/channel/election/mod.go:323]: you can add a return after

File: be1-go/channel/election/mod.go [hunk 6]
```diff
@@ -297,6 +306,53 @@ func (c *Channel) verifyMessage(msg message.Message) error {
...
+	err = senderPoint.UnmarshalBinary(senderBuf)
+	if err != nil {
+		return answer.NewError(-4, "invalid sender public key")
```
Comment (nkcr) [inline be1-go/channel/election/mod.go:326]:
```suggestion
		return answer.NewErrors(-4, "invalid sender public key: %s", senderBuf)
```

File: be1-go/channel/election/mod_test.go [hunk 7]
```diff
@@ -340,6 +341,93 @@ func Test_Cast_Vote_And_Gather_Result(t *testing.T) {
...
+	var pub method.Publish
+
+	err = json.Unmarshal(bufCreatePub, &pub)
+	require.NoError(t, err)
+
+	pub.Params.Message = m
+	pub.Params.Channel = electChannel.channelID
+
+	// publish the election open on the election channel
+	require.NoError(t, electChannel.Publish(pub, socket.ClientSocket{}))
```
Comment (nkcr) [inline be1-go/channel/election/mod_test.go:392]: would be clearer to do that in two steps:
```go
err := election...
require.NoError...
```

File: be1-go/channel/election/verification.go [hunk 8]
```diff
@@ -10,39 +10,104 @@ import (
...
+	}
+
+	// verify election id is base64URL encoded
+	_, err = base64.URLEncoding.DecodeString(electionOpen.Election)
+	if err != nil {
+		return xerrors.Errorf(elecIDBase64, electionOpen.Election)
+	}
+
+	// split channel to [lao id, election id]
+	noRoot := strings.ReplaceAll(c.channelID, messagedata.RootPrefix, "")
```
Comment (nkcr) [inline be1-go/channel/election/verification.go:37]: you can add a return after

File: be1-go/channel/election/verification.go [hunk 9]
```diff
@@ -10,39 +10,104 @@ import (
...
+	IDs := strings.Split(noRoot, "/")
+	if len(IDs) != 2 {
+		return xerrors.Errorf(elecIDFormat, c.channelID)
+	}
```
Comment (nkcr) [inline be1-go/channel/election/verification.go:41]: you can add a return after

File: be1-go/channel/lao/mod.go [hunk 10]
```diff
@@ -507,8 +508,7 @@ func (c *Channel) createElection(msg message.Message,
 	channelPath := "/root/" + setupMsg.Lao + "/" + setupMsg.ID
 
 	// Create the new election channel
-	electionCh := election.NewChannel(channelPath, setupMsg.StartTime, setupMsg.EndTime,
-		setupMsg.Questions, c.attendees, c.hub, c.log, c.organizerPubKey)
+	electionCh := election.NewChannel(channelPath, setupMsg, c.attendees, c.hub, c.log, c.organizerPubKey)
```
Comment (nkcr) [inline be1-go/channel/lao/mod.go:511]: sweet 😃

**Review by nkcr — COMMENTED** [review]
Not a lot to say, look very nice overall ! 💪 .
Just got minor comments.

**Review by nkcr — APPROVED** [review]

**Review by nkcr — APPROVED** [review]

**Pull request merged**